### PR TITLE
[SP-5809]Catch key error

### PR DESCRIPTION
**Pull request by eeerin**
https://losant.atlassian.net/browse/SP-5809

### CODE REVIEW

File: lib/model.js [hunk 1]
```diff
@@ -153,7 +162,16 @@ const _create = (data, definition) => {
       obj._etag = newEtag;
       if (explicitId) { obj._id = explicitId; }
       debug(`inserting into ${collection}: ${JSON.stringify(obj)}`);
-      const result = await db.insert(collection, obj);
+      try {
+        result = await db.insert(collection, obj);
+      } catch (err) {
+        if (err.message.startsWith('E11000 duplicate key') && definition.transformDuplicateKeyError) {
+          // this should throw an error.
+          definition.transformDuplicateKeyError(obj, err);
```
Comment (eeerin) [inline lib/model.js:170]: I thought about parsing the error message for parts like collection, index and key. 

https://rollbar.com/Losant/losant-workers/items/758/?item_page=0&item_count=100&#traceback

**Comment by eeerin** [comment]
I figured no one would ever want to actually use that function on a model like experienceUser.transformDuplicateKeyError() so why make it a method?

**Comment by spacetc62** [comment]
well, for the same reasons that before/after save, before/after remove are methods on the model - no one calls those directly either.

File: lib/model.js [hunk 2]
```diff
@@ -135,12 +135,21 @@ const _create = (data, definition) => {
...
       const query = { _id: obj._id, _etag: obj._etag };
       obj._etag = newEtag;
       debug(`updating ${collection}: ${JSON.stringify(obj)}`);
-      const result = await db.replace(collection, query, obj);
+      try {
+        result = await db.replace(collection, query, obj);
+      } catch (err) {
+        if (err.message.startsWith('E11000 duplicate key') && context.transformDuplicateKeyError) {
+          // this should throw an error.
+          context.transformDuplicateKeyError(obj, err);
```
Comment (spacetc62) [inline lib/model.js:148]: 2 thoughts
1) I think we should just throw the result here (have the function return the new error)
2) you need to call it like `context.transformDuplicateKeyError.apply(obj, [err]);` (like beforeSave) - that way the `this` in the function is in fact the model object

**Pull request merged**